zephyr/lib/alloc.c

```diff
@@ -145,14 +145,89 @@ static bool is_l3_heap_pointer(void *ptr)
 	uintptr_t l3_heap_start = get_l3_heap_start();
 	uintptr_t l3_heap_end = l3_heap_start + get_l3_heap_size();
 
-	if (is_cached(ptr))
-		ptr = z_soc_uncached_ptr((__sparse_force void __sparse_cache *)ptr);
-
 	if ((POINTER_TO_UINT(ptr) >= l3_heap_start) && (POINTER_TO_UINT(ptr) < l3_heap_end))
 		return true;
 
 	return false;
 }
+
+/**
+ * Flush cached L3 heap management data.
+ */
+static inline void l3_heap_flush(struct k_heap *h)
+{
+	if (h == &l3_heap)
+		sys_cache_data_flush_and_invd_range((void *)get_l3_heap_start(),
+						    get_l3_heap_size());
+}
+
+/**
+ * Invalidate cached L3 heap management data.
+ */
+static inline void l3_heap_invalidate(struct k_heap *h)
+{
+	if (h == &l3_heap)
+		sys_cache_data_invd_range((void *)get_l3_heap_start(),
+					  get_l3_heap_size());
+}
+
+static void *l3_heap_alloc_aligned(struct k_heap *h, size_t min_align, size_t bytes)
+{
+	k_spinlock_key_t key;
+	void *ret;
+#if CONFIG_SYS_HEAP_RUNTIME_STATS && CONFIG_IPC_MAJOR_4
+	struct sys_memory_stats stats;
+#endif
+
+	key = k_spin_lock(&h->lock);
+	l3_heap_invalidate(h);
+	ret = sys_heap_aligned_alloc(&h->heap, min_align, bytes);
+	l3_heap_flush(h);
```

Comment:
and after allocation you write back caches (to make changes visible to other cores, that's logical) and then you discard cached data again - why?

Reply:
Yes, this was mistake. Fixed.